### PR TITLE
bpo-45747: Detect gdbm/dbm dependencies in configure (GH-29467)

### PR DESCRIPTION
**Pull request by tiran**
Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue45747

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -3115,6 +3115,73 @@ else
...
+  AC_CHECK_LIB([gdbm], [gdbm_open])
+  LIBS=$LIBS_SAVE
+])
+
+# check for _dbmmodule.c dependencies
+AC_CHECK_HEADERS([ndbm.h], [
+  LIBS_SAVE="$LIBS"
+  AC_CHECK_LIB([ndbm], [dbm_open])
+  LIBS="$LIBS_SAVE"
+  AC_CHECK_LIB([gdbm_compat], [dbm_open])
```
Comment (erlend-aasland) [inline configure.ac:3130]: I wonder if we can fold these into:

```suggestion
  AC_SEARCH_LIBS([dbm_open], [ndbm gdbm_compat])
```

File: configure.ac [hunk 2]
```diff
@@ -3115,6 +3115,73 @@ else
...
+  LIBS="$LIBS_SAVE"
+])
+
+# "gdbm-ndbm.h" and "gdbm/ndbm.h" are both normalized to "gdbm_ndbm_h"
+AS_UNSET([ac_cv_header_gdbm_ndbm_h])
```
Comment (erlend-aasland) [inline configure.ac:3136]: Why do we need to unset `ac_cv_header_gdbm_ndbm_h` all the time?

File: configure.ac [hunk 3]
```diff
@@ -3115,6 +3115,73 @@ else
...
+# db.h re-defines the name of the function
+AC_CHECK_HEADERS([db.h], [
+  AC_CACHE_CHECK([for libdb], [ac_cv_have_libdb], [
+    LIBS_SAVE="$LIBS"
+    LIBS="$LIBS -ldb"
+    AC_LINK_IFELSE([AC_LANG_PROGRAM([
+      #define DB_DBM_HSEARCH 1
+      #include <db.h>
+      #if DB_VERSION_MAJOR < 5
+        #error "db.h < 5.x is not supported."
```
Comment (erlend-aasland) [inline configure.ac]: Will this error message be visible if this fails? If yes, maybe make the error message a bit more informative (if no, just mark this as resolved :):

```suggestion
        #error "dh.h: DB_VERSION_MAJOR < 5 is not supported."
```

File: configure.ac [hunk 4]
```diff
@@ -3115,6 +3115,73 @@ else
...
+      #endif
+      ], [DBM *dbm = dbm_open(NULL, 0, 0)])
+    ], [ac_cv_have_libdb=yes], [ac_cv_have_libdb=no])
+    LIBS="$LIBS_SAVE"
+  ])
+  AS_VAR_IF([ac_cv_have_libdb], [yes], [
+    AC_DEFINE(HAVE_LIBDB, 1, [Define to 1 if you have the `db' library (-ldb).])
```
Comment (erlend-aasland) [inline configure.ac]: For consistency

```suggestion
    AC_DEFINE([HAVE_LIBDB], [1], [Define to 1 if you have the `db' library (-ldb).])
```

**Review by erlend-aasland — COMMENTED** [review]
I left some comments about `configure.ac` (I haven't looked at the `setup.py` changes yet)

File: configure.ac (hunk 1 above)
Comment (tiran) [reply configure.ac:3130]: ``AC_SEARCH_LIBS()`` does not create ``HAVE_LIBFOO`` entries.

File: configure.ac (hunk 2 above)
Comment (tiran) [reply configure.ac:3136]: ``AC_CHECK_HEADER`` always caches the result in ``ac_cv_header_$name``. Both ``gdbm-ndbm.h`` and ``gdbm/ndbm.h`` are normalized to ``gdbm_ndbm_h``. It's super annoying but it's the only way to check for both header files without a false positive cache entry.

File: configure.ac (hunk 3 above)
Comment (tiran) [reply configure.ac]: I'm not sure, let's try!

File: configure.ac (hunk 1 above)
Comment (erlend-aasland) [reply configure.ac:3130]: Ah, that's unfortunate.

File: configure.ac (hunk 2 above)
Comment (erlend-aasland) [reply configure.ac:3136]: All right. Could you make the comment a little bit more explicit about that? :)

**Review by erlend-aasland — APPROVED** [review]
LGTM. Is it worth it to document the `USE_*` defines in `whatsnew`?

**Comment by tiran** [comment]
> LGTM. Is it worth it to document the `USE_*` defines in `whatsnew`?

It's an internal detailed of the build system and not user-facing.

**Pull request merged**